File: pipecat/lib/growthbook.py
"""GrowthBook feature flag integration.

Provides a shared async client that initializes once at startup and resolves
flags per-call using UserContext. Falls back gracefully when GrowthBook is
unavailable — all flags return their defaults.

Usage:
    from lib.growthbook import init_growthbook, resolve_flags, is_on, get_value

    # At startup (main.py)
    await init_growthbook()

    # Per call (bot.py)
    flags = await resolve_flags(senior_id=senior_id, timezone=tz, call_type="check-in")
    session_state["_flags"] = flags

    # In service code
    if is_on("director_enabled", session_state):
        ...
"""

from __future__ import annotations

from loguru import logger

_client = None
_initialized = False


async def init_growthbook() -> bool:
    """Initialize the GrowthBook client. Call once at startup.

    Returns True if initialization succeeded, False otherwise.
    When False, all flag checks return their defaults (safe degradation).
    """
    global _client, _initialized

    from config import settings

    if not settings.growthbook_api_host or not settings.growthbook_client_key:
        logger.info("GrowthBook not configured (no GROWTHBOOK_API_HOST/CLIENT_KEY)")
        return False

    try:
        from growthbook import GrowthBookClient, Options

        _client = GrowthBookClient(
            Options(
                api_host=settings.growthbook_api_host,
                client_key=settings.growthbook_client_key,
                cache_ttl=300,  # 5-minute cache
            )
        )
        success = await _client.initialize()
        if success:
            _initialized = True
            logger.info("GrowthBook initialized")
        else:
            logger.warning("GrowthBook initialization returned False — using defaults")
        return success
    except Exception as e:
        logger.warning("GrowthBook init failed — using defaults: {err}", err=str(e))
        return False


async def close_growthbook() -> None:
    """Clean up the GrowthBook client. Call at shutdown."""
    global _client, _initialized
    if _client:
        try:
            await _client.close()
        except Exception:
            pass
        _client = None
        _initialized = False


async def resolve_flags(
    senior_id: str | None = None,
    timezone: str | None = None,
    call_type: str = "check-in",
) -> dict:
    """Resolve all feature flags for a call. Returns a dict of flag values.

    When GrowthBook is unavailable, returns all defaults (everything enabled).
    Store the result in session_state["_flags"] for the duration of the call.
    """
    defaults = {
        "director_enabled": True,
        "news_search_enabled": True,
        "memory_search_enabled": True,
        "tts_fallback": False,
        "tts_provider": "cartesia",  # "cartesia" or "elevenlabs"
        "context_cache_enabled": True,
        "post_call_analysis_enabled": True,
        "scheduler_call_stagger_ms": 5000,
    }

    if not _initialized or not _client:
        return defaults

    try:
        from growthbook import UserContext

        user = UserContext(
            attributes={
                "id": senior_id or "unknown",
                "timezone": timezone or "UTC",
                "call_type": call_type,
            }
        )

        resolved = {}
        for key, default in defaults.items():
            if isinstance(default, bool):
                resolved[key] = await _client.is_on(key, user)
            else:
                resolved[key] = await _client.get_feature_value(key, default, user)
        return resolved
    except Exception as e:
        logger.warning("GrowthBook flag resolution failed — using defaults: {err}", err=str(e))
        return defaults


def is_on(flag: str, session_state: dict, default: bool = True) -> bool:
    """Check if a flag is enabled from session_state["_flags"].

    Synchronous — reads from the pre-resolved flags dict.
    Returns default if flags haven't been resolved yet.
    """
    flags = session_state.get("_flags", {})
    return flags.get(flag, default)


def get_value(flag: str, session_state: dict, default=None):
    """Get a flag value from session_state["_flags"].

    Synchronous — reads from the pre-resolved flags dict.
    """
    flags = session_state.get("_flags", {})
    return flags.get(flag, default)

File: lib/growthbook.js
/**
 * GrowthBook feature flag integration for Node.js.
 *
 * Shared client that initializes once at startup. Resolves flags per
 * scheduler cycle using user attributes. Falls back gracefully when
 * GrowthBook is unavailable — all flags return their defaults.
 *
 * Usage:
 *   import { initGrowthBook, closeGrowthBook, resolveFlags, isOn, getValue } from './lib/growthbook.js';
 *
 *   // At startup (index.js)
 *   await initGrowthBook();
 *
 *   // Per scheduler cycle
 *   const flags = await resolveFlags({ id: seniorId, timezone });
 *   const stagger = getValue('scheduler_call_stagger_ms', flags, 5000);
 */

import { GrowthBook } from '@growthbook/growthbook';
import { createLogger } from './logger.js';

const log = createLogger('GrowthBook');

let client = null;
let initialized = false;

const DEFAULTS = {
  director_enabled: true,
  news_search_enabled: true,
  memory_search_enabled: true,
  tts_fallback: false,
  context_cache_enabled: true,
  post_call_analysis_enabled: true,
  scheduler_call_stagger_ms: 5000,
};

export async function initGrowthBook() {
  const apiHost = process.env.GROWTHBOOK_API_HOST;
  const clientKey = process.env.GROWTHBOOK_CLIENT_KEY;

  if (!apiHost || !clientKey) {
    log.info('GrowthBook not configured (no GROWTHBOOK_API_HOST/CLIENT_KEY)');
    return false;
  }

  try {
    client = new GrowthBook({
      apiHost,
      clientKey,
      enableDevMode: process.env.NODE_ENV !== 'production',
    });
    await client.loadFeatures({ timeout: 5000 });
    initialized = true;
    log.info('GrowthBook initialized');
    return true;
  } catch (err) {
    log.warn('GrowthBook init failed — using defaults', { error: err.message });
    return false;
  }
}

export function closeGrowthBook() {
  if (client) {
    client.destroy();
    client = null;
    initialized = false;
  }
}

export async function resolveFlags(attributes = {}) {
  if (!initialized || !client) {
    return { ...DEFAULTS };
  }

  try {
    client.setAttributes(attributes);
    const resolved = {};
    for (const [key, defaultVal] of Object.entries(DEFAULTS)) {
      if (typeof defaultVal === 'boolean') {
        resolved[key] = client.isOn(key);
      } else {
        resolved[key] = client.getFeatureValue(key, defaultVal);
      }
    }
    return resolved;
  } catch (err) {
    log.warn('GrowthBook flag resolution failed — using defaults', { error: err.message });
    return { ...DEFAULTS };
  }
}

export function isOn(flag, flags, defaultVal = true) {
  if (!flags) return defaultVal;
  return flags[flag] ?? defaultVal;
}

export function getValue(flag, flags, defaultVal = null) {
  if (!flags) return defaultVal;
  return flags[flag] ?? defaultVal;
}
